src/catalog/src/durable/impls/migrate.rs

```diff
@@ -0,0 +1,227 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use async_trait::async_trait;
+use mz_ore::now::EpochMillis;
+
+use crate::durable::debug::{DebugCatalogState, Trace};
+use crate::durable::impls::persist::UnopenedPersistCatalogState;
+use crate::durable::impls::stash::OpenableConnection;
+use crate::durable::{
+    BootstrapArgs, CatalogError, DurableCatalogState, OpenableDurableCatalogState,
+};
+
+// Note: All reads done in this file can be fenced out by a new writer. All writers start by first
+// fencing the stash and then fencing persist. Therefore, when performing operations like:
+//
+//   1. Read tombstone from stash.
+//   2. Use tombstone to decide to either read something from the stash or persist.
+//
+// We don't have to worry about race conditions where a new writer updates the stash or persist
+// inbetween steps (1) and (2).
+
+#[derive(Debug)]
+pub(crate) enum Direction {
+    MigrateToPersist,
+    RollbackToStash,
+}
+
+#[derive(Debug)]
+pub struct CatalogMigrator {
+    openable_stash: Box<OpenableConnection>,
+    openable_persist: Box<UnopenedPersistCatalogState>,
+    direction: Direction,
+}
+
+#[async_trait]
+impl OpenableDurableCatalogState for CatalogMigrator {
+    async fn open_savepoint(
```

Comment:
Good point, I think the readonly function should panic since there's no valid uses. This one I think is useful though, it will allow us to use the upgrade checker tool to check that a migration would succeed.

Reply:
What about the open writeable method? Should that one also panic? Agree it's useful for the upgrade checker.

Reply:
The `open` method is what is going to actually execute the migration/rollback, so I think we need to leave it as is.